midend/predication.h

```diff
@@ -23,51 +23,89 @@ limitations under the License.
-branches, but in this way we cannot have two side-effects in the same
-conditional statement.
-
 */
 class Predication final : public Transform {
-    NameGenerator* generator;
+    /**
+     * Private Visitor only for Predication pass.
+     * This pass checks nested Mux expressions(?:)
+     * in assignment statement. 
+     * It checks if the right assignment expression
+     * can be placed in the nested Mux DAG.
+     */
+    class ReplaceChecker final : public Inspector {
+        const IR::AssignmentStatement * assignmentStatement;
+        bool conflictDetected = false;
+        unsigned currentNestingLevel = 0;
+        const std::vector<bool>& travesalPath;
+     public:
+        explicit ReplaceChecker(std::vector<bool>& t)
+        : travesalPath(t)
+        { }
+        const bool& isConflictDetected() const { return conflictDetected; }
+        bool preorder(const IR::Mux * mux) override;
+        bool preorder(const IR::AssignmentStatement * statement) override;
+    };
+
+    class EmptyStatementRemover final : public Transform {
+     public:
+        EmptyStatementRemover() {}
+        IR::BlockStatement* preorder(IR::BlockStatement * block) override;
+    };
+
+    /**
+     * Private Transformer only for Predication pass.
+     * This pass operates on nested Mux expressions(?:). 
+     * It replaces then and else expressions in Mux with 
+     * the appropriate expression from assignment.
+     */
+    class ExpressionReplacer final : public Transform {
+     private:
+        IR::AssignmentStatement * statement;
+        const IR::Expression * rightExpression;
+        const std::vector<bool>& travesalPath;
+        const std::vector<const IR::Expression *>& conditions;
+        unsigned currentNestingLevel = 0;
+     public:
+        explicit ExpressionReplacer(const IR::Expression * e,
+                                    std::vector<bool>& t,
+                                    std::vector<const IR::Expression *>& c)
+        : rightExpression(e), travesalPath(t), conditions(c)
+        { CHECK_NULL(e); }
+        const IR::Mux * preorder(IR::Mux * mux) override;
+        const IR::AssignmentStatement * preorder(IR::AssignmentStatement * statement) override;
+        void emplaceExpression(IR::Mux * mux);
+        void visitThen(IR::Mux * mux);
+        void visitElse(IR::Mux * mux);
+    };
+
+    EmptyStatementRemover remover;
+    IR::BlockStatement * currentBlock;
     bool inside_action;
-    std::vector<cstring> predicateName;
     unsigned ifNestingLevel;
+    // Traverse path of nested if-else statements
+    // true at the end of the vector means that you are currently visiting 'then' branch'
+    // false at the end of the vector means that you are in the else branch of the if statement.
+    // Size of this vector is the current if nesting level.
+    std::vector<bool> travesalPath;
+    std::vector<const IR::Expression *> nestedConditions;
+    std::vector<cstring> orderedNames;
```

Comment:
Can a name appear more than once?  If not, it would be more efficient to use an `ordered_set` here.

Reply:
Name can't appear more than once, refactored this to the _ordered_set_.